sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -69,6 +72,38 @@ class Analyzer(
   )
 
   /**
+   * Substitute CTE definitions
+   */
+  object CTESubstitution extends Rule[LogicalPlan] {
+    def apply(plan: LogicalPlan): LogicalPlan = {
+      val (realPlan, cteRelations) = plan match {
```

Comment:
Do we still need to first get `realPlan` and `creRelations`? Can we just use the following? 
```
def apply(plan: LogicalPlan): LogicalPlan = plan match {
  case With(child, relations) => substituteCTE(child, relations)
  case other => other 
}
```